Stop sending CloneSessionStorageNamespace message from web process

szewai opened the pull request:
#### 3533fb3edb8034827a24eeeb72feb65c247b06b0
<pre>
Stop sending CloneSessionStorageNamespace message from web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=273082">https://bugs.webkit.org/show_bug.cgi?id=273082</a>
<a href="https://rdar.apple.com/126879290">rdar://126879290</a>

Reviewed by Alex Christensen.

Currently web process sends CloneSessionStorageNamespace message to network process when creating new window (page). On
receiving the message, network process copies SessionStorage of the existing page to new page. Since web process cannot
be trusted, network process is supposed to validate that the sender process has access to both pages before performing
the copy, but it currently doesn&apos;t. To fix this, this patch moves the message from web process to UI process. UI process
is a trusted process, so network process does not need to do the validation. Also, UI process is reponsible for creating
new page (see WebPageProxy::CreateNewPage), so it knows what pages need the copy.

This is for hardening and it should not change existing behavior. The SessionStorage copy behavior is covered by
existing tests like storage/domstorage/sessionstorage/window-open.html and
imported/w3c/web-platform-tests/webstorage/storage_session_window_noopener.window.html.

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createWindow):
* Source/WebCore/page/Page.h:
* Source/WebCore/storage/StorageNamespaceProvider.h:
(WebCore::StorageNamespaceProvider::cloneSessionStorageNamespaceForPage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::cloneSessionStorageNamespace):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace): Deleted.
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::cloneSessionStorageNamespaceForPage):
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace): Deleted.
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createWindow):

Canonical link: <a href="https://commits.webkit.org/277988@main">https://commits.webkit.org/277988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46d51d694ab807cdd8983c9db73f7c6c7205d024

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51737 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45102 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51336 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25781 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40079 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50239 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42275 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21187 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23355 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43449 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7135 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45266 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53653 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24088 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20343 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47395 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42483 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46365 "Found 5 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/selection/listbox, /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/selection/menulist, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/value/basic, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10811 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26158 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25082 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/storage/StorageNamespaceProvider.h:
```diff
@@ -55,7 +55,7 @@ class StorageNamespaceProvider : public RefCounted<StorageNamespaceProvider> {
 
     void setSessionStorageQuota(unsigned quota) { m_sessionStorageQuota = quota; }
-    virtual void copySessionStorageNamespace(Page&, Page&) = 0;
+    virtual void cloneSessionStorageNamespaceForPage(Page&, Page&) { RELEASE_ASSERT_NOT_REACHED(); }
```
I don't see a good reason why this can't be pure virtual.

szewai replied:
After the change, only WebKitLegacy::WebStorageNamespaceProvider should provide the implementation for the function and the function should not be invoked on other inherited class object, so it seems more convenient to add the implementation here.

achristensen07 commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -772,11 +772,20 @@ void NetworkStorageManager::cloneSessionStorageForWebPage(WebPageProxyIdentifier
     ASSERT(!m_closed);
 
     m_queue->dispatch([this, protectedThis = Ref { *this }, fromIdentifier, toIdentifier]() mutable {
-        assertIsCurrent(workQueue());
```
We may as well keep this even though it's now redundant.

achristensen07 commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -772,11 +772,20 @@ void NetworkStorageManager::cloneSessionStorageForWebPage(WebPageProxyIdentifier
+{
+    assertIsCurrent(workQueue());
+
+    for (auto& [key, manager] : m_originStorageManagers) {
```
Key is unused.  Why not just iterate .values()?

szewai committed:
Stop sending CloneSessionStorageNamespace message from web process
https://bugs.webkit.org/show_bug.cgi?id=273082
rdar://126879290

Reviewed by Alex Christensen.

Currently web process sends CloneSessionStorageNamespace message to network process when creating new window (page). On
receiving the message, network process copies SessionStorage of the existing page to new page. Since web process cannot
be trusted, network process is supposed to validate that the sender process has access to both pages before performing
the copy, but it currently doesn't. To fix this, this patch moves the message from web process to UI process. UI process
is a trusted process, so network process does not need to do the validation. Also, UI process is reponsible for creating
new page (see WebPageProxy::CreateNewPage), so it knows what pages need the copy.

This is for hardening and it should not change existing behavior. The SessionStorage copy behavior is covered by
existing tests like storage/domstorage/sessionstorage/window-open.html and
imported/w3c/web-platform-tests/webstorage/storage_session_window_noopener.window.html.

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createWindow):
* Source/WebCore/page/Page.h:
* Source/WebCore/storage/StorageNamespaceProvider.h:
(WebCore::StorageNamespaceProvider::cloneSessionStorageNamespaceForPage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::cloneSessionStorageNamespace):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace): Deleted.
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::cloneSessionStorageNamespaceForPage):
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace): Deleted.
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createWindow):

Canonical link: https://commits.webkit.org/277988@main
